nix/hsie/Main.hs

```diff
@@ -0,0 +1,349 @@
+    , impQualified  :: ImportQualified
+    , impAs         :: Maybe Text
+    , impHiding     :: ImportHiding
+    , impSymbol     :: Maybe Text
+    }
+    deriving (Generic, Csv.ToNamedRecord, Csv.DefaultOrdered, Aeson.ToJSON)
+
+data ImportQualified
+  = Qualified
+  | NotQualified
+  deriving (Eq, Generic, Aeson.ToJSON)
+
+instance Csv.ToField ImportQualified where
+  toField Qualified    = "qualified"
+  toField NotQualified = "not qualified"
+
+data ImportHiding
+  = Wildcard
+  | Hiding
+  | NotHiding
+  deriving (Eq, Generic, Aeson.ToJSON)
+
+instance Csv.ToField ImportHiding where
+  toField Wildcard  = "wildcard"
+  toField Hiding    = "hiding"
+  toField NotHiding = "not hiding"
+
+-- | Mapping of modules to their aliases and to the files they are found in
+type ModuleAliases = [(Text, [(Text, [Text])])]
+
+-- | Mapping of modules to files
+type WildcardImports = [(Text, [Text])]
+
+
+-- MAIN
+
+main :: IO ()
+main =
+  run =<< O.customExecParser prefs infoOpts
+  where
+    prefs = O.prefs $ O.subparserInline <> O.showHelpOnEmpty
+    infoOpts =
+      O.info (O.helper <*> opts) $
+        O.fullDesc
+        <> O.header "hsie - Swiss army knife for HaSkell Imports and Exports"
+        <> O.progDesc "Parse Haskell code to analyze it's imports and exports"
+    opts =
+      Options <$> commandOption <*> O.some (srcOption <|> importsOption)
+    srcOption =
+      O.option (SourceDir <$> O.str) $
+        O.long "src"
+        <> O.short 's'
+        <> O.metavar "SRCDIR"
+        <> O.help "Haskell source directory"
+        <> O.action "directory"
+    importsOption =
+      O.option (ImportsDir <$> O.str) $
+        O.long "imports"
+        <> O.short 'i'
+        <> O.metavar "IMPORTSDIR"
+        <> O.help "Imports directory generated with GHC's '--ddump-minimal-imports'"
+        <> O.action "directory"
+    commandOption =
+      O.subparser $
+        command "dump" "Dump imported symbols as CSV or JSON"
+          (Dump <$> jsonOutputFlag)
+        <> command "graph-modules" "Print dot graph of module imports"
+             (pure GraphModules)
+        <> command "graph-symbols" "Print dot graph of symbol imports"
+             (pure GraphSymbols)
+        <> command "check-aliases"
+             "Check that aliases of imported modules are consistent"
+             (pure CheckAliases)
+        <> command "check-wildcards"
+             "Check that no unwanted modules are imported as wildcards"
+             (CheckWildcards <$> O.many okModuleOption)
+    command name desc options =
+      O.command name . O.info (O.helper <*> options) $ O.progDesc desc
+    jsonOutputFlag =
+      O.flag OutputCsv OutputJson $
+        O.long "json" <> O.short 'j' <> O.help "Output JSON"
+    okModuleOption =
+      O.option O.auto $
+        O.long "ok"
+        <> O.short 'o'
+        <> O.metavar "OKMODULE"
+        <> O.help "Module that is ok to import as wildcard"
+
+run :: Options -> IO ()
+run Options{..} =
```

Comment:
In this case, where the fields don't have a prefix, I find it much more readable to use `NamedFieldPuns` instead of `RecordWildcards`:

```suggestion
run Options{command, sources} =
```